compiler/src/iree/compiler/Dialect/Flow/Transforms/FormScalarDispatches.cpp

```diff
@@ -197,7 +197,18 @@ void FormScalarDispatchesPass::runOnOperation() {
       }
 
       if (!isSliceRoot(scalarWorkloadLimit, prevOp)) {
-        if (isClonableIntoDispatchOp(prevOp)) {
+        // Keep looking for more roots to horizontally fuse if the `prevOp` has
+        // no dependencies. e.g. `tensor.empty()`
+        if (prevOp->getNumOperands() == 0 && prevOp->getNumRegions() == 0)
+          continue;
+
+        // If `prevOp` is clonable AND only used by the fused/cloned then it
+        // will get cloned before any possible roots that may define its
+        // operands. (if it is not used by the region it may not be cloned)
+        if (isClonableIntoDispatchOp(prevOp) &&
```

Comment:
Maybe we just stop forming dispatches past clonable ops. They will get transitively cloned

Reply:
I think this would result in nearly no horizontal fusion since any root directly proceeding a `tensor.empty` would not be fusible with any other op. For example, in this test case https://github.com/iree-org/iree/blob/c28823103cf86d971ee46e5405d9badeacaa286d/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/form_scalar_dispatches.mlir#L97-L116

`%2` and `%6` can't be placed in the same dispatch but because `%5 = tensor.empty() : tensor<f32>` is between them.

Reply:
Ok, i see your point... let me think about this a bit more.